rama-http/src/layer/har/spec.rs

```diff
@@ -0,0 +1,231 @@
+
+        Self {
+            method: parts.method.to_string(),
+            url: parts.uri.to_string(),
+            http_version: Self::into_string_version(parts.version),
+            cookies: vec![],
+            headers: Self::into_har_headers(&parts),
+            query_string: Self::into_har_query_string(&parts),
+            post_data: None,
+            headers_size: 0,
+            body_size: -1,
+            comment: None,
+        }
+    }
+
+    // this needs to be refactored somewhere else as
+    // it's widely used across the codebase
+    fn into_string_version(v: HttpVersion) -> String {
+        match v {
+            HttpVersion::HTTP_09 => String::from("0.9"),
+            HttpVersion::HTTP_10 => String::from("1.0"),
+            HttpVersion::HTTP_11 => String::from("1.1"),
+            HttpVersion::HTTP_2 => String::from("2"),
+            HttpVersion::HTTP_3 => String::from("3"),
+            _ => panic!("WTF!!"),
+        }
+    }
+
+    fn into_har_query_string(parts: &ReqParts) -> Vec<QueryString> {
+        parts
+            .uri
+            .query()
+            .map(|qs| {
+                qs.split('&')
+                    .filter_map(|kv| {
+                        let mut split = kv.split('=');
+                        let (name, value) =
+                            (split.next().unwrap_or(""), split.next().unwrap_or(""));
+                        Some(QueryString {
+                            name: name.to_string(),
+                            value: value.to_string(),
+                            comment: None,
+                        })
+                    })
+                    .collect::<Vec<_>>()
+            })
+            .unwrap_or_default()
+    }
+
+    fn into_har_headers(parts: &ReqParts) -> Vec<Header> {
+        parts
+            .headers
+            .clone()
+            .into_iter()
+            .map(|(name, value)| Header {
+                name: name.unwrap().to_string(),
+                value: value.to_str().unwrap_or_default().to_string(),
+                comment: None,
+            })
+            .collect::<Vec<_>>()
+    }
+}
+
+har_data!(Response, {
+    pub status: u16,
+    pub status_text: String,
+    pub http_version: String,
+    pub cookies: Vec<Cookie>,
+    pub headers: Vec<Header>,
+    pub content: Content,
+    pub redirect_url: String,
+    pub headers_size: i64,
+    pub body_size: i64,
+    pub comment: Option<String>,
+});
+
+har_data!(Cookie, {
+    pub name: String,
+    pub value: String,
+    pub path: Option<String>,
+    pub domain: Option<String>,
+    pub expires: Option<String>,
+    pub http_only: Option<bool>,
+    pub secure: Option<bool>,
+    pub comment: Option<String>,
+});
+
+har_data!(Header, {
+    pub name: String,
+    pub value: String,
+    pub comment: Option<String>,
+});
+
+har_data!(QueryString, {
+    pub name: String,
+    pub value: String,
+    pub comment: Option<String>,
+});
+
+har_data!(PostData, {
+    pub mime_type: String,
```

Comment:
any stuff like this which comes from header data, try to use the typed headers as much as possible as they do a lot of heavy lifting for you where they exist. E.g. Mime type from the `ContentType` header.

You find all available ones at https://ramaproxy.org/docs/rama/http/headers/index.html

Reply:
> and I'm not sure whether I can do it using some other utils already included

What do you mean?

Reply:
if there is some code I can reuse somewhere in the library

Reply:
tbh it's because I was thinking of using some external crate to represent har (when i'll be finished with filling up the fields), which might help for future serialization/deserialization and conversion to other file formats. For the same reason it should keep primitive types: if at some point you'd like to export to file, it's much less laborious to have primitive types which can be more easily converted to string. In general it feels very right to keep the most general types, especially for interoperability, present and future.

Reply:
Well that's where I do not agree. I prefer to keep. the dependencies as little as possible. No need to depend on anything else for Har. And we can already provide a Recorder impl for `File` async. Others can impl their own thing. The serialization can also already be handled by primitives provided by this har modules that can be used across Recorders if desired.

But no we do not want to turn everything into a `String` just to then serialize it anyway. That's gives a lot of allocations for no benefit.

Reply:
Ok, I added Mime and (!) Query. Is it what you envisioned?

Reply:
`Mime` is a primitive / final type. E.g. similar to `Host`, or `Authority`, `Domain`, etc...
`Query` is a wrapper type, a utility. It is only there to "extract". Hence why it is in the `extract` modules.